### ISSUE TITLE
Do not show tooltips after closing dialogs

### ISSUE THREAD

**Issue by diegohaz** [reakit/reakit#750]
### Motivation

The reason for this question is better expressed via a short screen recording:

![CleanShot 2020-09-23 at 23 15 07](https://user-images.githubusercontent.com/15199/94093060-adeb1880-fdf2-11ea-9dd6-0a2f3cd2e1d9.gif)

I open a menu. However, when I click outside the menu the focus is automatically set back to the `MenuButton`. This then triggers the tooltip to appear, as it is usually expected when a button with a tooltip receives focus.

However, I'd argue that in this case the tooltip appearance is a nuisance. This means that all menu buttons with a tooltip will have their tooltip appear when the menu is dismissed. Even clicking a menu option will leave a tooltip floating around. This is presumably true for all buttons that trigger a popover or even a dialog.

I first wanted to know if this is a reasonable expectation on my side, and if so, then I'd like to discuss solutions. I was hoping this is something I could fix in user-land without the library having to fix it for me.

_Originally posted by @gnapse in https://github.com/reakit/reakit/discussions/749_

### Possible implementations

> I first wanted to know if this is a reasonable expectation on my side, and if so, then I'd like to discuss solutions. I was hoping this is something I could fix in user-land without the library having to fix it for me.

I think that is a really reasonable expectation.

As for the solution:

I think both this and tooltip delays (that we're experimenting with the `unstable_timeout` option) are part of the same feature: intent. We should set heuristics to determine whether the user has intention to see the tooltip.

With mouse over, delaying the tooltip for some milliseconds seems to do a pretty good job.

But I still don't know how to approach it for keyboard focus. We can't simply use "programmatic focus" as the criteria. For example, moving focus between items in a toolbar is technically a programmatic focus, but we do want it to trigger the tooltip.

We could maybe pass some metada to the element together with `element.focus()` when we don't want to show the tooltip. For example, when the dialog closes:
https://github.com/reakit/reakit/blob/c4c57cf43f22c8769dd439b0c989af4af79e1a83/packages/reakit/src/Dialog/__utils/useFocusOnHide.ts#L43

But, ideally, the solution should be scoped to the `Tooltip` component so it would work with dialogs from other libraries as well. I haven't tested it, but I guess something like this may work in userland already:
```jsx
<TooltipReference
  onFocus={(event) => {
    // If focus is not followed by a key up event, does it mean that it's not
    // an intentional keyboard focus? Not sure.
    const handleKeyUp = (e) => e.key !== "Escape" && tooltip.show();
    event.currentTarget.addEventListener("keyup", handleKeyUp, { once: true });
    // Prevent tooltip.show from being called by TooltipReference
    event.preventDefault();
  }}
/>
```

_Originally posted by @diegohaz in https://github.com/reakit/reakit/discussions/749#discussioncomment-83997_

**Comment by tom-sherman** [comment]
Something that I've been doing that I first saw in [Braid](https://github.com/seek-oss/braid-design-system) is to globally track if the user is a keyboard user. I've been doing this to only show focus styles on buttons for mouse users, as soon as the user is detected to be navigating by the keyboard we show the focus rings. This can be seen in Braid in their [`hideFocusRings` module](https://github.com/seek-oss/braid-design-system/blob/d7abbfe63266e48061a966c43518539abce82d1d/lib/components/private/hideFocusRings/useHideFocusRings.ts).

This to me is the same idea of "intent", and could be worth considering here. (Even if the idea of hiding focus rings for mouse users is potentially worrisome from an a11y perspective 😅)

One other thing I'd mention is that I'm not sure if any heuristic can fit everyone's use cases for this.

**Comment by gnapse** [comment]
This issue seems to be back in the new `@ariakit/react` (meaning, the given workaround no longer works). I believe part of the reason why it no longer works is that the tooltip anchor now uses `onFocusVisible`, right?

Is it planned to fix this in the library itself? Or maybe it is possible to achieve with an adaptation of the workaround shared above? I tried a few things, but no luck.

**Comment by diegohaz** [comment]
Hey @gnapse! I think it's better to open a new issue.

**Comment by FlorianBurgevin** [comment]
@gnapse i'm facing the same problem. Did you open a new issue ?

**Comment by gnapse** [comment]
I did not. Mostly because I realized that the issue I'm seeing now is not the same I reported here originally. The one reported here can be summed up like this (see bolded part):

> In a menu button with a tooltip, **when closing the open menu with the mouse by clicking outside**, the menu button receives focus and the tooltip appears.

However, with the new Ariakit (`@ariakit/react`) I only see the tooltip appear if the menu is closed via keyboard. I find this more acceptable. And the previous issue (closing the menu by clicking outside) does not trigger the tooltip.

However, do open a new issue if you think it should never show the tooltip.
